src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1548,20 +1529,52 @@ void DoCancelRequest(int backendProcessId, int backendSecretKey)
+            UserCancellationToken = cancellationToken;
+            ReadBuffer.Cts.ResetCts();
+
+            AttemptPostgresCancellation = attemptPgCancellation;
+            return _cancellationTokenRegistration =
+                cancellationToken.Register(static c => ((NpgsqlConnector)c!).PerformUserCancellation(), this);
+        }
+
         /// <summary>
-        /// Resets cancellation-related state to prepare for a new cancellable operation.
+        /// Starts a new cancellable operation within an ongoing user action. This should only be used if a single user action spans several
+        /// different actions which each has its own cancellation tokens. For example, a command execution is a single user action, but
+        /// spans ExecuteReaderQuery, NextResult, Read and so forth.
         /// </summary>
-        /// <param name="userCancellationToken">
-        /// The cancellation token provided by the user for the operation. This is stored on the connector, and will be referenced
-        /// by the <see cref="OperationCanceledException"/> if one is thrown.
+        /// <param name="cancellationToken">
+        /// The cancellation token provided by the user. Callbacks will be registered on this token for executing the cancellation,
+        /// and the token will be included in any thrown <see cref="OperationCanceledException"/>.
+        /// </param>
+        /// <param name="attemptPgCancellation">
+        /// If <see langword="true" />, PostgreSQL cancellation will be attempted when the user requests cancellation or a timeout
+        /// occurs, followed by a client-side socket cancellation once <see cref="NpgsqlConnectionStringBuilder.CancellationTimeout"/> has
+        /// elapsed. If <see langword="false" />, PostgreSQL cancellation will be skipped and client-socket cancellation will occur
+        /// immediately.
         /// </param>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal void ResetCancellation(CancellationToken userCancellationToken)
+        internal CancellationTokenRegistration StartNestedCancellableOperation(
+            CancellationToken cancellationToken = default,
+            bool attemptPgCancellation = true)
         {
-            _cancellationRequested = false;
-            _userCancellationRequested = false;
-            UserCancellationToken = userCancellationToken;
-            ReadBuffer.Cts.ResetCts();
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // NpgsqlDataReader.Consume calls NextResult, but is also called from within NextResult - so we need to support reentrancy.
+            // If a non-empty registration already exists, the user's token was already registered up the stack, so we do nothing.
+            if (_cancellationTokenRegistration != default)
+                return default;
+
+            UserCancellationToken = cancellationToken;
+            AttemptPostgresCancellation = attemptPgCancellation;
```

Comment:
This is a little bit dangerous, because it's possible to have a nested cancellable operation set this flag to `false`, and after dispose we're still working like we're in an outer scope (I have not been able to find something like this yet).